sys/drv/bcm2835_emmc.c

```diff
@@ -0,0 +1,535 @@
+      break;
+    case EMMC_PROP_RW_BLKSIZE:
+      if (var >= 1024)
+        return EINVAL;
+      reg = b_in(emmc, BCMEMMC_BLKSIZECNT) & ~BSC_BLKSIZE;
+      b_out(emmc, BCMEMMC_BLKSIZECNT, reg | var);
+      break;
+    case EMMC_PROP_RW_RESP_LOW:
+      b_out(emmc, BCMEMMC_RESP0, (uint32_t)var);
+      b_out(emmc, BCMEMMC_RESP1, (uint32_t)(var >> 32));
+      break;
+    case EMMC_PROP_RW_RESP_HI:
+      b_out(emmc, BCMEMMC_RESP2, (uint32_t)var);
+      b_out(emmc, BCMEMMC_RESP3, (uint32_t)(var >> 32));
+      break;
+    case EMMC_PROP_RW_CLOCK_FREQ:
+      return bcmemmc_set_clk_freq(cdev->parent, var);
+    case EMMC_PROP_RW_BUSWIDTH:
+      return bcmemmc_set_bus_width(state, var);
+    case EMMC_PROP_RW_RCA:
+      state->rca = var;
+      break;
+    default:
+      return EINVAL;
+  }
+
+  return 0;
+}
+
+static int bcmemmc_send_cmd(device_t *cdev, emmc_cmd_t cmd, uint32_t arg,
+                            emmc_resp_t *resp) {
+  device_t *dev = cdev->parent;
+  bcmemmc_state_t *state = (bcmemmc_state_t *)dev->state;
+  resource_t *emmc = state->emmc;
+  uint32_t code = bcmemmc_encode_cmd(cmd);
+
+  if (cmd.flags & EMMC_F_APP)
+    bcmemmc_send_cmd(cdev, EMMC_CMD(APP_CMD), state->rca << 16, NULL);
+
+  b_out(emmc, BCMEMMC_ARG1, arg);
+  b_out(emmc, BCMEMMC_CMDTM, code);
+
+  int error = bcmemmc_intr_wait(dev, INT_CMD_DONE);
+  if (error) {
+    klog("ERROR: failed to send EMMC command %p", code);
+    return error;
+  }
+
+  if (resp) {
+    resp->r[0] = b_in(emmc, BCMEMMC_RESP0);
+    resp->r[1] = b_in(emmc, BCMEMMC_RESP1);
+    resp->r[2] = b_in(emmc, BCMEMMC_RESP2);
+    resp->r[3] = b_in(emmc, BCMEMMC_RESP3);
+  }
+
+  return 0;
+}
+
+static int bcmemmc_read(device_t *cdev, void *buf, size_t len, size_t *donep) {
+  device_t *emmcdev = cdev->parent;
+  bcmemmc_state_t *state = (bcmemmc_state_t *)emmcdev->state;
+  resource_t *emmc = state->emmc;
+  uint32_t *data = buf;
+
+  assert(is_aligned(len, sizeof(uint32_t)));
+
+  /* TODO(morhcore): should use DMA transfer if possible. */
+  for (size_t i = 0; i < len / sizeof(uint32_t); i++)
+    data[i] = b_in(emmc, BCMEMMC_DATA);
+
+  /* TODO(mohrcore): check if the transfer fully succeeded! */
+  if (donep)
+    *donep = len;
+  return 0;
+}
+
+static int bcmemmc_write(device_t *cdev, const void *buf, size_t len,
+                         size_t *donep) {
+  device_t *emmcdev = cdev->parent;
+  bcmemmc_state_t *state = (bcmemmc_state_t *)emmcdev->state;
+  resource_t *emmc = state->emmc;
+  const uint32_t *data = buf;
+
+  assert(is_aligned(len, sizeof(uint32_t)));
+
+  /* TODO(morhcore): should use DMA transfer if possible. */
+  for (size_t i = 0; i < len / sizeof(uint32_t); i++)
+    b_out(emmc, BCMEMMC_DATA, data[i]);
+
+  /* TODO(mohrcore): check wether the transfer fully succeeded! */
+  if (donep)
+    *donep = len;
+  return 0;
+}
+
+/* e.MMC requires some GPIO setup to work properly. This however is different
```

Comment:
The reason is very simple. Please just look at https://www.raspberrypi.org/app/uploads/2012/04/Raspberry-Pi-Schematics-R1.0.pdf (page 2). Please modify the comment appropriately.

Reply:
done